runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/MetricAggregator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.spark.metrics;
+
+import java.io.Serializable;
+import org.apache.beam.sdk.metrics.DistributionData;
+import org.apache.beam.sdk.metrics.MetricKey;
+
+
+/**
+ * Metric values wrapper which adds aggregation methods.
+ * @param <ValueT> Metric value type.
+ */
+abstract class MetricAggregator<ValueT> implements Serializable {
```

Comment:
Ah I see. I'm a bit worried because (I believe) those represent implementation details of the Direct runner. Specifically, it includes notions of how the `DirectRunner` does threading, how to lock to make the DirectRunner perform reasonably, and supports getting the final result out. Within most other runners (such as Dataflow) I would expect that a different implementation might be more appropriate.

For instance, in Dataflow we don't use a MetricAggregator at all -- we report the metrics to a service (or output them to an accumulator/aggergator) and let those send them to a single worker which then does the aggregation. Thus there is no need for locking and stuff.

In the DirectRunner there is no ability to send them somewhere else for aggregation so it has to do the locking.

I'm not sure we should be sharing details between the DirectRunner (designed to be correct, but not efficient) with other runners (designed to be efficient and correct).